src/Microsoft.AspNetCore.Blazor.Razor.Extensions/ComponentDocumentRewritePass.cs

```diff
@@ -218,48 +217,55 @@ private void RewriteChildren(RazorSourceDocument source, IntermediateNode node)
                                             stack.Pop();
                                         }
 
-                                        if (token.Type == HtmlTokenType.EndTag)
+                                        break;
+                                    }
+
+                                case HtmlTokenType.EndTag:
+                                    {
+                                        var tag = token.AsTag();
+
+                                        var popped = stack.Pop();
+                                        if (stack.Count == 0)
                                         {
-                                            var popped = stack.Pop();
-                                            if (stack.Count == 0)
+                                            // If we managed to 'bottom out' the stack then we have an unbalanced end tag.
+                                            // Put back the current node so we don't crash.
+                                            stack.Push(popped);
+
+                                            var tagName = parser.GetTagNameOriginalCasing(token.AsTag());
```

Comment:
Can `tag` be used here instead of calling `AsTag()` again?

Reply:
yeah, you're right 👍